File: crypto/golden_test.go
package crypto

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// goldenFile is the shape of testdata/golden.json — amadan's pinned
// cross-implementation fixture (fixed keypairs, contexts, plaintexts and
// their expected outputs). This file is byte-for-byte amadan's
// internal/envelope/testdata/golden.json; passing it is what licenses
// amadan (and keymail, and seapointish) to delete their local copies and
// import this package instead.
type goldenFile struct {
	Envelope struct {
		Keypair json.RawMessage `json:"keypair"`
		SignPub string          `json:"sign_pub"`
		BoxPub  string          `json:"box_pub"`
		Seal    struct {
			Context   string `json:"context"`
			Plaintext string `json:"plaintext"`
			Sealed    string `json:"sealed"`
		} `json:"seal"`
		Sign struct {
			Context   string `json:"context"`
			Message   string `json:"message"`
			Signature string `json:"signature"`
		} `json:"sign"`
	} `json:"envelope"`
	Repokey struct {
		Key    string `json:"key"`
		Derive struct {
			Pack     string `json:"pack"`
			Manifest string `json:"manifest"`
			Browse   string `json:"browse"`
			CI       string `json:"ci"`
		} `json:"derive"`
		SealSym struct {
			Plaintext string `json:"plaintext"`
			Sealed    string `json:"sealed"`
		} `json:"sealsym"`
	} `json:"repokey"`
}

func loadGolden(t *testing.T) goldenFile {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("testdata", "golden.json"))
	if err != nil {
		t.Fatalf("read testdata/golden.json: %v", err)
	}
	var g goldenFile
	if err := json.Unmarshal(raw, &g); err != nil {
		t.Fatalf("unmarshal testdata/golden.json: %v", err)
	}
	return g
}

func mustB64(t *testing.T, s string) []byte {
	t.Helper()
	b, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		t.Fatalf("base64 decode %q: %v", s, err)
	}
	return b
}

// TestGoldenVectors replays amadan's pinned fixture against this
// implementation — the compatibility contract the crypto prompt names.
func TestGoldenVectors(t *testing.T) {
	g := loadGolden(t)

	t.Run("keypair unmarshal matches recorded public keys", func(t *testing.T) {
		kp, err := UnmarshalKeypair(g.Envelope.Keypair)
		if err != nil {
			t.Fatalf("UnmarshalKeypair: %v", err)
		}
		if !bytes.Equal(kp.SignPub(), mustB64(t, g.Envelope.SignPub)) {
			t.Fatal("SignPub derived from pinned keypair JSON does not match pinned sign_pub")
		}
		if !bytes.Equal(kp.BoxPub(), mustB64(t, g.Envelope.BoxPub)) {
			t.Fatal("BoxPub derived from pinned keypair JSON does not match pinned box_pub")
		}
	})

	t.Run("Open replays the pinned sealed blob", func(t *testing.T) {
		kp, err := UnmarshalKeypair(g.Envelope.Keypair)
		if err != nil {
			t.Fatalf("UnmarshalKeypair: %v", err)
		}
		wantPlain := mustB64(t, g.Envelope.Seal.Plaintext)
		sealed := mustB64(t, g.Envelope.Seal.Sealed)

		got, err := Open(kp, g.Envelope.Seal.Context, sealed)
		if err != nil {
			t.Fatalf("Open: %v", err)
		}
		if !bytes.Equal(got, wantPlain) {
			t.Fatalf("Open(pinned sealed) = %q, want %q", got, wantPlain)
		}
	})

	t.Run("Verify replays the pinned signature", func(t *testing.T) {
		msg := mustB64(t, g.Envelope.Sign.Message)
		sig := mustB64(t, g.Envelope.Sign.Signature)
		signPub := mustB64(t, g.Envelope.SignPub)

		if !Verify(signPub, g.Envelope.Sign.Context, msg, sig) {
			t.Fatal("Verify rejected the pinned golden signature")
		}
	})

	t.Run("Derive reproduces the pinned subkeys", func(t *testing.T) {
		key := mustB64(t, g.Repokey.Key)

		cases := []struct {
			ctx  string
			want string
		}{
			{"amadan-pack-v1", g.Repokey.Derive.Pack},
			{"amadan-manifest-v1", g.Repokey.Derive.Manifest},
			{"amadan-browse-v1", g.Repokey.Derive.Browse},
			{"amadan-ci-v1", g.Repokey.Derive.CI},
		}
		for _, c := range cases {
			got := Derive(key, c.ctx)
			want := mustB64(t, c.want)
			if !bytes.Equal(got, want) {
				t.Errorf("Derive(key, %q) = %x, want %x", c.ctx, got, want)
			}
		}
	})

	t.Run("OpenSym replays the pinned sealed blob", func(t *testing.T) {
		key := mustB64(t, g.Repokey.Key)
		wantPlain := mustB64(t, g.Repokey.SealSym.Plaintext)
		sealed := mustB64(t, g.Repokey.SealSym.Sealed)

		got, err := OpenSym(key, sealed)
		if err != nil {
			t.Fatalf("OpenSym: %v", err)
		}
		if !bytes.Equal(got, wantPlain) {
			t.Fatalf("OpenSym(pinned sealed) = %q, want %q", got, wantPlain)
		}
	})
}

File: crypto/js/golden.test.mjs
// The JS twin's half of the compatibility contract: replay the same
// pinned fixture (../testdata/golden.json) the Go package replays, plus
// cross-shape round trips. Run via `node --test` (Node ≥ 20); the Go
// test suite runs it automatically when node is on PATH (js_test.go).
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import {
  importKeypair, generate, seal, open, sign, verify,
  derive, sealSym, openSym, unb64, b64,
  deriveInvite, newInviteSecret, wrapKey, unwrapKey, newKey,
} from "./crypto.mjs";

const golden = JSON.parse(
  await readFile(fileURLToPath(new URL("../testdata/golden.json", import.meta.url)), "utf8"));

const kp = importKeypair(
  golden.envelope.keypair,
  unb64(golden.envelope.sign_pub),
  unb64(golden.envelope.box_pub));

test("open replays the pinned sealed blob", async () => {
  const got = await open(kp, golden.envelope.seal.context, unb64(golden.envelope.seal.sealed));
  assert.equal(b64(got), golden.envelope.seal.plaintext);
});

test("verify replays the pinned signature", async () => {
  const ok = await verify(
    unb64(golden.envelope.sign_pub),
    golden.envelope.sign.context,
    unb64(golden.envelope.sign.message),
    unb64(golden.envelope.sign.signature));
  assert.equal(ok, true);
});

test("verify rejects the pinned signature under the wrong context", async () => {
  const ok = await verify(
    unb64(golden.envelope.sign_pub),
    "not-the-context",
    unb64(golden.envelope.sign.message),
    unb64(golden.envelope.sign.signature));
  assert.equal(ok, false);
});

test("derive reproduces the pinned subkeys", async () => {
  const key = unb64(golden.repokey.key);
  for (const [ctx, want] of [
    ["amadan-pack-v1", golden.repokey.derive.pack],
    ["amadan-manifest-v1", golden.repokey.derive.manifest],
    ["amadan-browse-v1", golden.repokey.derive.browse],
    ["amadan-ci-v1", golden.repokey.derive.ci],
  ]) {
    assert.equal(b64(await derive(key, ctx)), want, `derive(${ctx})`);
  }
});

test("openSym replays the pinned sealed blob", async () => {
  const got = await openSym(unb64(golden.repokey.key), unb64(golden.repokey.sealsym.sealed));
  assert.equal(b64(got), golden.repokey.sealsym.plaintext);
});

test("seal here, open with the golden private key", async () => {
  const plain = new TextEncoder().encode("js twin round trip");
  const sealed = await seal(kp.boxPub, "twin-v1", plain);
  const got = await open(kp, "twin-v1", sealed);
  assert.deepEqual(got, plain);
});

test("sign here verifies here, and fails on tamper", async () => {
  const fresh = await generate();
  const msg = new TextEncoder().encode("attest");
  const sig = await sign(fresh, "attest-v1", msg);
  assert.equal(sig.length, 64);
  assert.equal(await verify(fresh.signPub, "attest-v1", msg, sig), true);
  msg[0] ^= 1;
  assert.equal(await verify(fresh.signPub, "attest-v1", msg, sig), false);
});

test("sealSym/openSym round trip with a fresh key", async () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const plain = new TextEncoder().encode("symmetric");
  const got = await openSym(key, await sealSym(key, plain));
  assert.deepEqual(got, plain);
});

const invites = JSON.parse(
  await readFile(fileURLToPath(new URL("../testdata/invites.json", import.meta.url)), "utf8"));

// unb64url: the vectors carry unpadded base64url; unb64 wants standard.
function unb64url(s) {
  let t = s.replace(/-/g, "+").replace(/_/g, "/");
  while (t.length % 4 !== 0) t += "=";
  return unb64(t);
}

test("deriveInvite replays eleven's pinned vectors", async () => {
  for (const vec of invites.invites) {
    const inv = await deriveInvite(invites.context, vec.t);
    assert.equal(inv.id, vec.id);
    assert.equal(inv.claimSecret, vec.claimSecret);
    assert.equal(inv.claimHash, vec.claimHash);
    // The pinned wrapped blob opens under the derived wrapKey to the
    // pinned group key (wrap uses a random IV, so only unwrap replays).
    const key = await unwrapKey(inv.wrapKey, vec.wrappedGroupKey);
    assert.deepEqual(key, unb64url(vec.groupKey));
  }
});

test("wrapKey/unwrapKey round trip, foreign context refused", async () => {
  const secret = newInviteSecret();
  const inv = await deriveInvite("app-invite", secret);
  const key = newKey();
  const wrapped = await wrapKey(inv.wrapKey, key);
  assert.deepEqual(await unwrapKey(inv.wrapKey, wrapped), key);

  const other = await deriveInvite("other-app", secret);
  await assert.rejects(() => unwrapKey(other.wrapKey, wrapped));
});
